ARCs/arc-0056.md

```diff
@@ -317,31 +317,34 @@ interface SourceInfo {
 
 ### Template Variables
 
-Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to either `pushint` or `pusbyte` (not the pseudo opcodes `int` or `byte`). Using these opcodes is important to ensure that the program does not add them to the constant blocks.
+Template variables are variables in the TEAL that should be substitued prior to compilation. Template variables **MUST** be an argument to `pusbytes` (not the pseudo opcode `byte`) to ensure that the program does not add them to the constant blocks. `pushint` (or `int`) is not used due to the variance in integer byte sizes which impact source mapping.
```

Comment:
My main thinking was that it adds a lot of complexity to substitution. If we allow TMPL_ variables in `bytecblock` or `pushbytess` you have more complexity due to it being mixed in with other value, which may also contain `TMPL_` in non-template variable values.

Reply:
I think the regex would still be pretty simple though?

Reply:
I’m sure I’ve seen (PyTEAL?) contracts that have TMPL in constant block so it’s something that is in the wild…

Reply:
I think if we don't prohibit `TMPL_` values appearing in constant blocks then that provides the opportunity for some other interesting solutions when debugging traces.

For example we can produce a debug map for a high level language that is independent of actual template values. By using pc offsets from the first op after the constant blocks, then when paired with a trace file we can determine the true pc values.

Reply:
Regarding the regex, would it also need to include negative matches for other values inside the quote? e.g. don't match `"some strange TMPL_VALUE string"`

Reply:
> I think if we don't prohibit TMPL_ values appearing in constant blocks then that provides the opportunity for some other interesting solutions when debugging traces.

Ah yeah that is interesting. That is why I emit `dissasembledTeal` in TEALScript, so you can use the dissasemlbed TEAL as the ultimate source of truth. 

> Regarding the regex, would it also need to include negative matches for other values inside the quote? e.g. don't match "some strange TMPL_VALUE string"

Right we'd also need to account for escaping (and escaping backslashes) which adds a lot of complexity

Reply:
There's an existing regex that's been used in algokit utils for ages that's always seemed to work. It doesn't detect if the value is meant to be a uint vs bytes though (it does it based on whether it's passed a bigint/number or a uint8array.

Reply:
The current implementations don't properly support bytecblock. If there was a string in the constant block with `TMPL_` then there would be a false match. I suppose we could simply forbid anything but hex in the constant block (in fact now that I think about it I'm not even sure what is supported)

Reply:
Pretty sure the current algokit utils implementation supports bytecblock, that’s what all the unit tests work against funnily enough because pyteal uses it

Reply:
But the question is not just if it supports bytecblock (because of course a regex would still work on a bytecblock), but whether or not it will erroneously replace `TMPL_` strings that are not intended to be a template variable. 

For example `bytecblock TMPL_FOO "TMPL_FOO"`

The first `TMPL_FOO` should be replaced but not the second. It is certainly an edge case though and maybe we think it's accetable?

Reply:
And by acceptable I mean compilers don't let you use the string `TMPL_` anywhere. Otherwise the behavior would be unexpected.

Reply:
I agree and I think it's a bit of an edge case.

Funnily enough, someone had a problem like that in python and the python utils has more complex code to deal with it (but it never made its way to typescript and noone complained).

Reply:
Namely https://github.com/algorandfoundation/algokit-utils-py/blob/main/src/algokit_utils/deploy.py#L291

Reply:
In terms of ARC-56, it should be sufficient to say what is allowed, but it doesn't need to describe how that is achieved (i.e. regex or otherwise) as that is an implementation detail?

Reply:
Additionally I think we should still allow AVM uint64 template values. 

In cases where a fixed size uint value is required (i.e.  the cases\ described [here](https://github.com/algorandfoundation/ARCs/pull/258#issuecomment-2344642857)) then the template value can use one of the ARC4 numeric types to inform how a value should be encoded?

Reply:
Wholeheartedly agree on both points.

Reply:
iirc the Python code handles comments but not same-line strings. I could be wrong though. Either way it's a decent amount of complexity which is why I think it's important to outline a reference algorithim in the ARC. 

> Additionally I think we should still allow AVM uint64 template values.
>
> In cases where a fixed size uint value is required (i.e. the case described 
https://github.com/algorandfoundation/ARCs/pull/258#issuecomment-2344642857) then the template value can use one of the ARC4 numeric types to inform how a value should be encoded?

If we can agree on how source mapping should be done with them I agree, otherwise I don't think giving up human-readable errors is worth uint64 template variables. 

The options I see for source mapping are

A) Include decompliled TEAL lines in source map
B) Include bytecblock offset in source map

Both would require calls to algod awhich is not great and can be avoided (for integers at least) if we stick to byte-only, but it'd still be a problem for dynamic types. I think B is preferable because it doesn't depend on stability from the dissasembler

Reply:
Now we have a refined approach to template variables we should change the parts of this ARC that refer to needing to use pushbytes